Add support for text in WebGL renderers

jahow opened the pull request:
This PR adds text rendering in WebGLVector and WebGLVectorTile renderers.

A specific "Text Overlay" worker is introduced: this worker takes in Render Instructions generated from Features, and renders them using the Canvas2D rendering pipeline (`TextBuilder` and `Executor`). It then gives back an Offscreen canvas with the text rendered, and this canvas is overlaid on every frame of the WebGL renderers. This approach tries to limit blocking the main thread as well as using transferable structures as much as possible.

<img width="1378" height="484" alt="image" src="https://github.com/user-attachments/assets/834c9a47-8fc9-4024-827e-dee5044d1f5a" />


Examples now using text:

https://deploy-preview-16693--ol-site.netlify.app/en/latest/examples/webgl-vector-layer.html
https://deploy-preview-16693--ol-site.netlify.app/en/latest/examples/webgl-vector-tiles.html
https://deploy-preview-16693--ol-site.netlify.app/en/latest/examples/webgl-points-layer.html

Work items:
- [x] Re-render map when text overlay becomes ready
- [x] Fix issue with non-ansi chars
- [x] Support pixel ratio > 1
- [x] Support rotated map view
- [x] Support text overflow computation
- [x] Add rendering tests

**Important:**

This feature comes with a few caveat, to be addressed later on:
* text repeat is not working great; also, automatic text positioning is sometimes off
* no support yet for style variables (`var` operator) in text styles
* no support yet for custom CSS fonts
* no decluttering yet
* no hit detection on labels yet

jahow commented:
I'd like to give a small update on this PR: I have encountered a lot of issues while trying to achieve good performance, especially with vector tiles. I will pick up work on this at some point but currently I've had to move to other priorities, unfortunately.

jahow commented:
Performance is much better now, as the text overlay worker is now directly handling canvas rendering instructions!

TheOnlyArtz commented:
> @TheOnlyArtz I'm reaching completion on this PR I think! the code is now close to being final, there's still a couple of small bugs to fix, tests to write and some more performance optimization.

Thank you very much, we are very excited to migrate @jahow

karussell commented:
This looks really great and impressive work! Looking forward to it :)

I tested the example "WebGL Vector Tiles Layer". One small bug I noticed is for e.g. [this hamlet](https://www.openstreetmap.org/node/338618001) - the name "Sollschwitz - Sulšecy" is rendered with a "?" at the end (latest desktop & mobile firefox & chromium)

<img width="381" height="187" alt="image" src="https://github.com/user-attachments/assets/56f49420-7acb-401a-aef4-a7cf2c338d41" />

However several other non-Ansi characters work properly. (Also it seems that only one line is supported for the name and the name for one line is limited to a certain length? leading to partial or incorrect name data.)

Furthermore there is a strange "offset" bug in latest mobile firefox (update April 2026: also appears now in desktop Firefox), where the text labels are moving at a different (double?) speed compared to the map when panning the map. Leading to incorrect offsets:

<img width="256" alt="image" src="https://github.com/user-attachments/assets/87fb0d89-b87b-4ac6-8c72-4dbd4d7ca5c5" />

jahow committed:
Add text to webgl examples

Also switch flat styles in the mapbox vector tiles examples (canvas and webgl)

jahow committed:
Add util for serializing & deserializing framestate objects

jahow committed:
Webgl / adapt compileUtil to support text attributes

We now use 3 floats to store texts in buffers instead of just one (
first is the integer equivalent of the text, second is the position in the
text array and third is the length in the text array); and
string values are not converted to int here anymore; this will be done
later on

jahow committed:
Introduce a LabelsArray class that serializes strings into transferable arrays of UTF8 bytes

jahow committed:
WebGL / introduce utilities for rendering text

These will generate canvas text builders out of render instructions, and
help setting up a webgl post process for rendering the text overlay on top
of other layers

jahow committed:
Webgl / adapt render instructions util to handle strings

jahow committed:
Webgl / move renderinstructions tests to node

jahow committed:
Add utils for mat4: reset, rotate

jahow committed:
Webgl / show compilation errors & support mat4 uniforms in PostProcessing

jahow committed:
Webgl / adapt style parsing to logic to also parse text style props

Also the attributes callbacks do not convert the string to an int right away,
it is down later on, so the tests reflect that.

jahow committed:
Webgl / Add text overlay worker

This worker takes in render instructions and draw the corresponding text
on an offscreen canvas that is then given back to the main thread
on every render loop

jahow committed:
WebGL / Adapt Vectorstylerenderer to maintain a canvas with the text overlay

The text overlay worker is used for rendering the labels on every frame; the
resulting render is kept in the textOverlayCanvas on the VectorStyleRenderer,
and can be used in the final compositing

A new property "sourceRule" was added when converting flat style objects to
shaders; this is useful so that the source flat style can be read back for
transferring to the text overlay worker

jahow committed:
Webgl / adapt vector layer renderer to support text rendering

jahow committed:
webgl / dispose text instructions when disposing of a vector tile representation

jahow committed:
Webgl Helper / avoid error when providing an empty postProcesses array

jahow committed:
Webgl / adapt vector tile layer renderer to support text rendering

jahow committed:
Style / add a more general flatStyleLikeToStyleFunction util

This is now used for canvas vector layers as well

jahow commented:
There's still a few tests failing and a few things left to do, but this PR is close to being ready for review. It will not change in significant ways anymore.

Note: I will create smaller PRs for some stuff that were added here to make it simpler to review overall, if we indeed agree on this approach for text rendering.

karussell commented:
Thanks a lot for your work here and that you continue investing your time :)

Not sure if it is important, but the strange text offset/moving issue, which I saw under Android, now appears in latest Firefox and Safari on latest macOS. (but is now fine under Linux for both Chrome and Firefox)

jahow commented:
Thank you! I will test the feature with other platforms. Are you sure it's not just that there are too many labels to render?

karussell commented:
It also happens for the other example with only a few labels: https://deploy-preview-16693--ol-site.netlify.app/en/latest/examples/webgl-vector-tiles.html

karussell commented:
https://github.com/user-attachments/assets/c672af8e-fb9d-4030-a7c2-db7a723efb1b

jahow commented:
Thank you, most likely a pixel ratio issue :)

jahow committed:
Webgl / use a common function for messaging workers

this fixes memory leaks because some listeners were never unsubscribed

fredj commented on src/ol/render/webgl/serialize.js:
```diff
@@ -0,0 +1,53 @@
+    size: frameState.size,
+    extent: frameState.extent,
+    coordinateToPixelTransform: frameState.coordinateToPixelTransform,
+    pixelToCoordinateTransform: frameState.coordinateToPixelTransform,
```
```suggestion
    pixelToCoordinateTransform: frameState.pixelToCoordinateTransform,
```

jahow replied:
this commit was amended, thanks

jahow committed:
WebGL text / do not clear the overlay if the render is incomplete (flashing)

jahow committed:
WebGL text / use view resolution as input when generating text instructions

This allows us to correctly compute polygon sizes and support the
`overflow: false` style parameter

jahow committed:
Webgl Vector / do not bake view rotation when computing buffer transform

This was causing the rotation of the text rendered in the textOverlay worker
to be wrong because the view rotation was applied twice

jahow committed:
adjust typescript strict config

jahow committed:
Add rendering tests for webgl text rendering

Those are mostly inspire from canvas rendering tests, with a few workarounds
for unsupported features

jahow committed:
WebGL / refresh layers once the text overlay is ready

jahow commented:
This is finally ready for review! I've added the various known limitations in the PR body.

fredj commented:
there's an issue in `stripNonTextStyleProperties` when providing a style array, can be reproduced with:
```diff
diff --git a/test/node/ol/render/webgl/textUtil.test.js b/test/node/ol/render/webgl/textUtil.test.js
index 7ee63e92f3..f66a47fb27 100644
--- a/test/node/ol/render/webgl/textUtil.test.js
+++ b/test/node/ol/render/webgl/textUtil.test.js
@@ -262,6 +262,28 @@ describe('ol/render/webgl/textUtil', function () {
         },
       ]);
     });
+    it('only keeps the style properties relevant to text rendering (rule with an array of styles)', function () {
+      expect(
+        stripNonTextStyleProperties([
+          {
+            style: [
+              {
+                'fill-color': 'red',
+                'text-value': 'foo',
+              },
+            ],
+          },
+        ]),
+      ).to.eql([
+        {
+          style: [
+            {
+              'text-value': 'foo',
+            },
+          ],
+        },
+      ]);
+    });
   });
 
   describe('createPostProcessDefinition', function () {
```

jahow committed:
WebGL text utils / Fix post review

jahow commented:
> there's an issue in stripNonTextStyleProperties when providing a style array, can be reproduced with:

Thank you, fixed in https://github.com/openlayers/openlayers/pull/16693/commits/8850c9c07e5fa69011de2cd56906b32dd91f6a68

fredj commented on src/ol/render/webgl/textUtil.js:
```diff
@@ -0,0 +1,482 @@
+export function stripNonTextStyleProperties(style) {
+  function stripStyle(style) {
+    for (const prop in style) {
+      if (!prop.startsWith('text-')) {
```
should we keep `z-index` style property here ?

jahow commented:
Thanks for the review @fredj, another reviewer with experience in WebGL would be very much appreciated for this before merging, thanks!

jahow commented:
Hi, I will wait for https://github.com/openlayers/openlayers/pull/17515 to be merged and then rebase this PR on top.

Once this is done, if everyone is ok with it (cc @openlayers/core) I plan on merging this PR. Thank you!

jahow committed:
chore: adapt tests for vitest & chai assert

jahow commented:
This is now rebased and uses vitest + chai for unit tests

ridoo commented:
@jahow That is great news to have text supported by WebGL styles now! Thanks for your work :)

> no support yet for style variables (var operator) in text styles

This seems to be also true for the `Rule.filter` property! This puzzles me as I would expect resolving the filter expression not being part of the documented caveat.

jahow commented:
> This seems to be also true for the Rule.filter property! This puzzles me as I would expect resolving the filter expression not being part of the documented caveat.

You're right, the caveat is probably not clear enough. The filter expression is also evaluated on the text overlay worker to know whether labels should be shown or not; this is done by the Canvas renderer, and it has the same limitations regarding `var`.

I'm eager to receive feedback on this feature, did you find out this limitation to be the biggest issue? How was performance on your side? Thanks!

ridoo commented:
@jahow Thanks for clarification :). I do have to migrate some layers which do have thousands of features to be rendered. The application is quite complex and uses lots of nested StyleFunctions. I have hoped there is a clear path I could follow (e.g. using vars for reactive properties). But this does not seem to be as straight-forward than I thought in the beginning : 

- defined flat.js-types are a good start, but the array-like syntax is quite cumbersome (some kind of DSL would be helpful to be honest)
- sometimes there is an issue comparing boolean, there's lots of recommendations to use 1/0 instead
- not all expressions are supported by WebGL drivers
- text styles are different to geometry styles
- no geometries for labels (lower priority)

Migrating the nested StyleFunctions is challenging by itself, unfortunately. Performance is to be tested (I will report back until I am back from vacation)

jahow commented:
Good luck in your migration, feel free to reach out for assistance (here or by email) if needed, thanks!
